File: backend/python/app/middlewares/error_middleware.py
from flask import jsonify, Flask
from werkzeug.exceptions import HTTPException
from app.utils.response_error import ResponseError
from pydantic import ValidationError


def error_middleware(app: Flask):
    @app.errorhandler(Exception)
    def handle_exception(error):
        app.logger.error(error)

        if isinstance(error, ValidationError):
            return (
                jsonify(
                    {
                        "errorMsg": "Request Not Valid!",
                        "errors": error.errors(),
                    }
                ),
                400,
            )

        if isinstance(error, ResponseError):
            if error.errors:
                return (
                    jsonify(
                        {
                            "errorMsg": error.args[0],
                            "errors": error.errors,
                        }
                    ),
                    error.status,
                )

            return (
                jsonify(
                    {
                        "errorMsg": error.args[0],
                    }
                ),
                error.status,
            )

        if isinstance(error, HTTPException):
            return (
                jsonify(
                    {
                        "errorMsg": error.description,
                    }
                ),
                error.code,
            )

        return (
            jsonify(
                {
                    "errorMsg": "Something went wrong. Check logs for detail errors!",
                }
            ),
            500,
        )

File: backend/nodejs-typescript-graphql/src/middlewares/error.middleware.ts
import { NextFunction, Request, Response } from "express";
import { ResponseError } from "@/utils/response-error.util";

export default async function errorMiddleware(
  error: Error,
  req: Request,
  res: Response,
  next: NextFunction,
) {
  if (!error) {
    next();
    return;
  }

  req.error = error;

  if (error instanceof ResponseError) {
    res.status(error.status).json({
      errorMsg: error.message,
      errors: error.errors,
    });
    return;
  }

  res.status(500).json({
    errorMsg: "Something went wrong. Check logs for detail errors!",
  });
}
